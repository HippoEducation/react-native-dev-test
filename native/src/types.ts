export type Post = {
  id: string;
  title: string;
  body: string;
  publishedAt: Date;
  author: {
    id: string;
    name: string;
  };
};

export type Posts = Post[];