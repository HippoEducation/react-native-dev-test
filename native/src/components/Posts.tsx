import React, { useState } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import {
  StyleSheet,
  Text,
  View,
  ActivityIndicator,
} from 'react-native';
import useGetLatinPostService from '../services/useGetPostService';
import PostList from './PostList';
import { Post } from '../types/Post';
import { RootStackParamList } from '../types/RootStackParamList';

type PostsScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Home'
>;

type Props = {
  navigation?: PostsScreenNavigationProp;
};

const Posts: React.FC<Props> = ({navigation}) => {
  const [posts, setPosts] = useState<Post[]>([]);
  
  const service = useGetLatinPostService(setPosts);

  const filterByAuthor = (name: string): void => setPosts(posts.filter((post) => post.author.name == name));

  const navigateToFullPost = (title: string, body: string): void => {
    navigation?.navigate('FullPost', { title: title, body: body });
  };

  return (
    <View style={styles.container}>
      {service.status === 'loading' && <ActivityIndicator />}
      {service.status === 'loaded' && <PostList 
        postList={posts}
        pressAuthorHandler={name => {filterByAuthor(name)}}
        pressPostHandler={(title, body) => navigateToFullPost(title, body)} 
      />}
      {service.status === 'error' && <Text>Oops, I did it again.</Text>}
    </View>
  );
};

export default Posts;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
});
