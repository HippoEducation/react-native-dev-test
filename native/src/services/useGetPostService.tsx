import { useEffect, useState } from 'react';
import { Post } from '../types/Post';
import { Service } from '../types/Service';

export interface Posts {
  posts: Post[];
}

const useGetPostService = () => {
  const [result, setResult] = useState<Service<Posts>>({
    status: 'loading',
  });

  useEffect(() => {
    fetch('http://10.0.0.122:4000/posts')
      .then((response) => response.json())
      .then((response) => setResult({ status: 'loaded', payload: response }))
      .catch((error) => setResult({ status: 'error', error }));
  }, []);

  return result;
};

export default useGetPostService;
