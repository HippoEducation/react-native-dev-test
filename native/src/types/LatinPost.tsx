export interface Author {
  id: string;
  name: string;
}

export interface LatinPost {
  author: Author;
  body: string;
  id: string;
  publishedAt: Date;
  title: string;
}
