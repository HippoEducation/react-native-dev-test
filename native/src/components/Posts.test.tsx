import React from 'react';
import renderer, { create, act } from 'react-test-renderer';
import { mocked } from 'ts-jest/utils';
import 'isomorphic-fetch';
import useGetPostService from '../services/useGetPostService';
import { Service } from '../types/Service';
import { Post } from '../types/Post';
import Posts from '../components/Posts';

jest.mock('../services/useGetPostService');
const mockedService = mocked(useGetPostService, true);

describe('<Posts />', () => {
  const postData = [
    {
      title: 'titular title',
      body: 'bod \n\n## stuff',
      author: {
        name: 'me',
        id: '1',
      },
      id: 'a',
      publishedAt: new Date('01-01-2020'),
    },
    {
      title: 'boring title',
      body: 'bod1 \n\n## stuff',
      author: {
        name: 'you',
        id: '2',
      },
      id: 'b',
      publishedAt: new Date(),
    },
  ];

  it('renders ActivityIndicator component while loading', () => {
    const expectedResult: Service<Post[]> = {
      status: 'loading',
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON().children[0];

    expect(tree.type).toBe('ActivityIndicator');
  });

  //TODO: find a way to actually assert on results
  it('displays post list reverse chronologically when loaded', () => {
    const expectedResult: Service<Post[]> = {
      status: 'loaded',
      payload: postData,
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON().children[0];

    expect(tree.type).not.toBe('ActivityIndicator');
  });

  it('displays error message when loaded with error', () => {
    const expectedResult: Service<Post[]> = {
      status: 'error',
      error: new Error(),
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON().children[0];

    expect(tree.type).not.toBe('ActivityIndicator');
    expect(tree.children[0]).toBe('Oops, I did it again.');
  });

  it('matches the snapshot', async () => {
    const expectedResult: Service<Post[]> = {
      status: 'loading',
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON();
    await act(async () => {expect(tree).toMatchSnapshot(); })
  });

  it('matches the snapshot', async () => {
    const expectedResult: Service<Post[]> = {
      status: 'error',
      error: new Error(),
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON();
    await act(async () => {expect(tree).toMatchSnapshot(); })
  });

  it('matches the snapshot', async () => {
    const expectedResult: Service<Post[]> = {
      status: 'loaded',
      payload: postData,
    };
    mockedService.mockReturnValueOnce(expectedResult);

    const tree = renderer.create(<Posts />).toJSON();
    await act(async () => {expect(tree).toMatchSnapshot(); })
  });
});
