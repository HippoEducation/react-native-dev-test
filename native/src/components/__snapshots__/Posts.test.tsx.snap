// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts /> matches the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`<Posts /> matches the snapshot 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Oops, I did it again.
  </Text>
</View>
`;

exports[`<Posts /> matches the snapshot 3`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#FFF",
          "fontWeight": "bold",
          "marginTop": 20,
        }
      }
    >
      Tap an author's name to only their posts.
    </Text>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "alignSelf": "center",
          "marginTop": 20,
          "width": "100%",
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
